### PR TITLE
feat: [#521] User can custom recover when a request panic

### LINKED ISSUE #521: ✨ [Feature] User can custom recover when a request panic
### Before feedback (在反馈之前)

- [X] There are no features that I want to submit in Issues (当前 Issues 中没有我将要提交的新功能)

### Describe Feature (描述功能)

Add a custom function to gin and fiber configurations.

### PR DESCRIPTION
**Pull request by KlassnayaAfrodita**
## 📑 Description

Closes https://github.com/goravel/goravel/issues/521

@coderabbitai summary

## ✅ Checks

- [ ] Added test cases for my code

### CODE REVIEW

**Comment by KlassnayaAfrodita** [comment]
I don't understand the reason for the error

File: route.go [hunk 1]
```diff
@@ -14,7 +14,7 @@ import (
 
 	"github.com/gofiber/fiber/v2"
 	"github.com/gofiber/fiber/v2/middleware/logger"
-	"github.com/gofiber/fiber/v2/middleware/recover"
+	fiberRecover "github.com/gofiber/fiber/v2/middleware/recover"
```
Comment (hwbrzzl) [inline route.go:17]:
```suggestion
	fiberrecover "github.com/gofiber/fiber/v2/middleware/recover"
```

**Review by hwbrzzl — COMMENTED** [review]
Thanks, LGTM, please optimize test cases.

File: middleware_timeout.go [hunk 2]
```diff
@@ -22,21 +22,14 @@ func Timeout(timeout time.Duration) contractshttp.Middleware {
...
-					LogFacade.Request(ctx.Request()).Error(r)
-					// TODO can be customized in https://github.com/goravel/goravel/issues/521
-					_ = ctx.Response().Status(http.StatusInternalServerError).String("Internal Server Error").Render()
-				}
-				close(done)
-			}()
-
+			defer HandleRecover(ctx, globalRecoverCallback)
 			ctx.Request().Next()
+			close(done)
```
Comment (hwbrzzl) [inline middleware_timeout.go:27]: The CI error is caused by this line, basically, the process will be blocked when moving this line here, gin has this issue as well, please create another PR to fix this: Keep this line in defer.

File: middleware_timeout.go [hunk 3]
```diff
@@ -22,21 +22,14 @@ func Timeout(timeout time.Duration) contractshttp.Middleware {
 		done := make(chan struct{})
 
 		go func() {
-			defer func() {
-				if r := recover(); r != nil {
-					LogFacade.Request(ctx.Request()).Error(r)
```
Comment (hwbrzzl) [inline middleware_timeout.go:27]: The log is required, please keep it. I saw it was removed in gin as well, please add it back.

**Comment by hwbrzzl** [comment]
Created a PR in gin to fix the issue: https://github.com/goravel/gin/pull/122, you can implement the same logic in fiber. Basically, the `HandleRecover` method can be removed.

File: middleware_timeout_test.go [hunk 4]
```diff
@@ -26,48 +25,66 @@ func TestTimeoutMiddleware(t *testing.T) {
...
+		require.NoError(t, err)
+
+		assert.Equal(t, http.StatusInternalServerError, resp.StatusCode)
+
+		body, err := io.ReadAll(resp.Body)
+		require.NoError(t, err)
+		assert.JSONEq(t, `{"error":"Internal Panic"}`, string(body))
+	})
+
+	mockLog.AssertExpectations(t)
```
Comment (hwbrzzl) [inline middleware_timeout_test.go:89]: It can be removed since we are using `mockslog.NewLog(t)`.

```suggestion
```

File: route.go [hunk 5]
```diff
@@ -130,6 +132,24 @@ func (r *Route) GlobalMiddleware(middlewares ...httpcontract.Middleware) {
...
+func (r *Route) Recover(callback func(ctx httpcontract.Context, err any)) {
+	globalRecoverCallback = callback
+	middleware := middlewaresToFiberHandlers([]httpcontract.Middleware{
+		func(ctx httpcontract.Context) {
+			defer func() {
+				if err := recover(); err != nil {
+					if callback != nil {
+						callback(ctx, err)
+					} else {
+						ctx.Request().AbortWithStatusJson(http.StatusInternalServerError, fiber.Map{"error": "Internal Server Error"})
```
Comment (hwbrzzl) [inline route.go:144]: The callback should not be `nil` here, so we can:

```suggestion
```

File: route_test.go [hunk 6]
```diff
@@ -21,6 +22,64 @@ import (
...
+	req := httptest.NewRequest("GET", "/recover", nil)
+	resp, err := route.Test(req)
+	assert.Nil(t, err)
+
+	body, err := io.ReadAll(resp.Body)
+	assert.Nil(t, err)
+	assert.Equal(t, "{\"error\":\"Internal Server Error\"}", string(body))
+	assert.Equal(t, http.StatusInternalServerError, resp.StatusCode)
+
+	mockConfig.AssertExpectations(t)
```
Comment (hwbrzzl) [inline route_test.go:48]:
```suggestion
```

File: route_test.go [hunk 7]
```diff
@@ -21,6 +22,64 @@ import (
...
+func TestRecoverWithDefaultCallback(t *testing.T) {
+	globalRecoverCallback = nil
+	mockConfig := configmocks.NewConfig(t)
+	mockConfig.On("GetBool", "http.drivers.fiber.prefork", false).Return(false).Once()
+	mockConfig.On("GetInt", "http.drivers.fiber.body_limit", 4096).Return(4096).Once()
+	mockConfig.On("GetInt", "http.drivers.fiber.header_limit", 4096).Return(4096).Once()
+
+	route, err := NewRoute(mockConfig, nil)
+	assert.Nil(t, err)
+	route.Recover(globalRecoverCallback)
```
Comment (hwbrzzl) [inline route_test.go:34]: This line is unnecessary in this case, we don't need to pass `nil` into the `Recover` method.

File: route_test.go [hunk 8]
```diff
@@ -21,6 +22,64 @@ import (
...
+	req := httptest.NewRequest("GET", "/recover", nil)
+	resp, err := route.Test(req)
+	assert.Nil(t, err)
+
+	body, err := io.ReadAll(resp.Body)
+	assert.Nil(t, err)
+	assert.Equal(t, "{\"error\":\"Internal Panic\"}", string(body))
+	assert.Equal(t, http.StatusInternalServerError, resp.StatusCode)
+
+	mockConfig.AssertExpectations(t)
```
Comment (hwbrzzl) [inline route_test.go:80]:
```suggestion
```

**Review by hwbrzzl — COMMENTED** [review]
Pretty good, a few nitpicks.

File: route_test.go (hunk 7 above)
Comment (KlassnayaAfrodita) [reply route_test.go:34]: in goravel/gin we used gin.Recovery() and the default recovery test covered this case. Is it necessary to test the default recoverer in goravel/fiber?

Comment (KlassnayaAfrodita) [reply route_test.go:34]: we can initially set globalRecoverGallback
```globalRecoverCallback := func(ctx contractshttp.Context, err any) {
		ctx.Request().AbortWithStatusJson(http.StatusInternalServerError, fiber.Map{"error": "Internal Server Error"})
	}```

Comment (hwbrzzl) [reply route_test.go]: There is default recovery as well:

```go
fiberHandlers := []fiber.Handler{
  fiberrecover.New(fiberrecover.Config{
	  EnableStackTrace: debug,
  }),
}
```

Setting a default `globalRecoverCallback`is a good idea.

**Review by hwbrzzl — APPROVED** [review]
LGTM

**Pull request merged**